File: Python/chapter 5/11724.py
import sys
sys.setrecursionlimit(10000)
input = sys.stdin.readline

node, edge = map(int, input().split())
graph = [[] for _ in range(node + 1)]
visited = [False] * (node+1)

def DFS(v):
    visited[v] = True
    for i in graph[v]:
        if not visited[i]:
            DFS(i)

for _ in range(edge):
    s, e = map(int, input().split())
    graph[s].append(e)
    graph[e].append(s)

count = 0

for i in range(1,node+1):
    if not visited[i]:
        DFS(i)
        count +=1

print(count)

File: CPP/chapter5/11724.cpp
#include <iostream>
#include <vector>

using namespace std;

static vector<vector<int>> A;
static vector<bool> visited;
void DFS(int v);

int main(){

    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int node, edge;
    cin >> node >> edge;

    A.resize(node+1);
    visited = vector<bool>(node+1, false);

    for(int i=0; i < edge ; i++){
        int n,m;
        cin >> n >> m ;
        A[n].push_back(m);
        A[m].push_back(n);
    }

    int count = 0;

    for(int i =1;i<node+1; i++){
        if(!visited[i]){
            count ++;
            DFS(i);
        }
    }

    cout << count << "\n";
}

void DFS(int v){
    if(visited[v]){
        return;
    }

    visited[v] = true;
    for(int i: A[v]){
        if(visited[i] == false){
            DFS(i);
        }
    }
}